### PR TITLE
Filter expansion

### PR DESCRIPTION
**Pull request by lukeatayde**
### Description:
This PR introduces a new mechanism for filtering out false positives.

In prior versions of Gitleaks, one could filter out results using a collection of regex patterns known as the "allowlist". Unfortunately, patterns in the allowlist only match on the secret itself, not the lines enclosing the entire secret. This means that filtering rules that needed to consider the surrounding context of a secret were impossible to express prior to this update.

Consider the following code snippet:
```
non_sensitive_key_1 = "<SOME_HASH_1>"
non_sensitive_key_2 = "<SOME_HASH_2>"
non_sensitive_key_3 = "<SOME_HASH_3>"
non_sensitive_key_4 = "<SOME_HASH_4>"
```
Previously, each of the keys would be caught by Gitleaks' generic key matching rule, but it would be impossible for the user to filter out the results. This is due to the fact that the allowlist regex patterns would only ever be supplied the `<SOME_HASH_i>` strings, instead of the entire line. As a result, a filter pattern like `non_sensitive_key_.*` wouldn't work to suppress the secrets.

To address this problem, the update introduces the ability to configure an "enclosing line allowlist". These allowlist match against _all_ the lines enclosed in a finding, not just the secret itself. As a result, one can filter out results like `non_sensitive_key_1 = "<SOME_HASH_1>"` using the pattern `non_sensitive_key_.*`. Like the older allowlists, these patterns can be applied globally, or to a particular rule.

While this update expands the list of patterns a user can filter out on, it isn't a complete solution. Unfortunately, Go's built-in regex library doesn't support positive or negative "lookaround" commands. This was an intentional decision by the language's developers to maintain linear time performance in Go's standard library. A developer with these assertions would have greater flexibility over what Gitleaks cab natively filter out. Unfortunately, the changes required to implement this would require a new library, refactoring, and a suspension of linear performance guarantees.

The modifications proposed in this MR could be simplified updating the older version of the allowlist to match on the entire finding instead of the secret alone. I decided to keep the old allowlist around to maintain backwards compatibility. Any patterns utilizing Regex anchor tokens would likely break otherwise.

### Other Changes

- Renamed the "line" field in the findings structure to "lines". This is to reflect the fact that the field can hold more than one line when it encodes newline characters.

### Checklist:

* [✅] Does your PR pass tests?
* [✅] Have you written new tests for your changes?
* [✅] Have you lint your code locally prior to submission?

### CODE REVIEW

**Review by Kennysan — APPROVED** [review]

File: config/config.go [hunk 1]
```diff
@@ -79,14 +82,10 @@ func (vc *ViperConfig) Translate() (Config, error) {
...
-		var allowlistRegexes []*regexp.Regexp
-		for _, a := range r.Allowlist.Regexes {
-			allowlistRegexes = append(allowlistRegexes, regexp.MustCompile(a))
-		}
-		var allowlistPaths []*regexp.Regexp
-		for _, a := range r.Allowlist.Paths {
-			allowlistPaths = append(allowlistPaths, regexp.MustCompile(a))
-		}
+
+		var allowlistRegexes = compileRegexPatterns(r.Allowlist.Regexes)
```
Comment (blevz) [inline config/config.go]: Prefer 
variableName := rhs 
Only use var when you don't yet have a value for the variable

File: report/finding.go [hunk 2]
```diff
@@ -13,7 +13,7 @@ type Finding struct {
 	StartColumn int
 	EndColumn   int
 
-	Line string `json:"-"`
+	Lines string `json:"-"`
```
Comment (blevz) [inline report/finding.go:16]: Should this change?

**Review by blevz — CHANGES_REQUESTED** [review]

File: config/config.go (hunk 1 above)
Comment (lukeatayde) [reply config/config.go]: Updated the changes to use := over var ... =

File: report/finding.go (hunk 2 above)
Comment (lukeatayde) [reply report/finding.go:16]: I intentionally changed the name to better reflect what the struct holds. When it finds a match, the field contains all of the lines the match was found on, not just the first line.

**Review by blevz — APPROVED** [review]
lgtm

**Pull request merged**